src/main/java/org/folio/services/invoice/InvoiceFiscalYearsService.java

```diff
@@ -0,0 +1,124 @@
+      String invoiceId, RequestContext requestContext) {
+    List<String> fiscalYearIds = budgets.stream()
+      .map(Budget::getFiscalYearId)
+      .distinct()
+      .collect(toList());
+    Map<String, List<String>> fundIdsByFiscalYearId = fiscalYearIds.stream()
+      .collect(toMap(
+        identity(),
+        fiscalYearId -> budgets.stream()
+          .filter(budget -> fiscalYearId.equals(budget.getFiscalYearId()))
+          .map(Budget::getFundId)
+          .distinct()
+          .collect(toList())
+      ));
+    List<String> possibleFiscalYearIds = fiscalYearIds.stream()
+      .filter(fiscalYearId -> fundIdsByFiscalYearId.get(fiscalYearId).size() == fundIds.size())
+      .collect(toList());
+    if (possibleFiscalYearIds.isEmpty()) {
+      log.error("getFiscalYearsByFundIdsAndBudgets(): no fiscal year matching all funds in invoice {}", invoiceId);
+      throwCouldNotFindValidFiscalYearError(invoiceId, fundIds);
+    }
+    String queryIds = convertIdsToCqlQuery(possibleFiscalYearIds);
+    LocalDate now = Instant.now().atOffset(ZoneOffset.UTC).toLocalDate();
+    String queryDate = "periodStart<=\"" + now + "\" sortby periodStart/sort.descending";
+    String query = String.format("%s AND %s", queryIds, queryDate);
+    return fiscalYearService.getFiscalYearCollectionByQuery(query, requestContext)
+      .map(fyCollection -> {
+        if (fyCollection.getTotalRecords() == 0) {
+          log.error("getFiscalYearsByFundIdsAndBudgets(): no fiscal year left after filtering by date in invoice {}",
+            invoiceId);
+          throwCouldNotFindValidFiscalYearError(invoiceId, fundIds);
+        }
+        checkUniqueSeries(invoiceId, fyCollection.getFiscalYears());
+        return fyCollection;
+      });
+  }
+
+  private void throwCouldNotFindValidFiscalYearError(String invoiceId, List<String> fundIds) {
+    List<Parameter> parameters = List.of(new Parameter().withKey("invoiceId").withValue(invoiceId),
+      new Parameter().withKey("fundIds").withValue(fundIds.toString()));
+    Error error = COULD_NOT_FIND_VALID_FISCAL_YEAR.toError()
+      .withParameters(parameters);
+    if (log.isErrorEnabled()) {
+      log.error("{} - {}", error.getMessage(),
+        parameters.stream().map(p -> String.format("%s=%s", p.getKey(), p.getValue())).collect(joining(", ")));
+    }
+    throw new HttpException(422, error);
+  }
+
+  private void checkUniqueSeries(String invoiceId, List<FiscalYear> fiscalYears) {
+    List<String> series = fiscalYears.stream().map(FiscalYear::getSeries).distinct().collect(toList());
+    if (series.size() <= 1)
+      return;
+    List<Parameter> parameters = List.of(new Parameter().withKey("invoiceId").withValue(invoiceId),
+      new Parameter().withKey("series").withValue(series.toString()));
+    Error error = MORE_THAN_ONE_FISCAL_YEAR_SERIES.toError()
+      .withParameters(parameters);
+    if (log.isErrorEnabled()) {
```

Comment:
It is unlikely. I added this test to avoid a code warning. A warning is generated when any value is computed in a call to one of the log methods (as an optimization step).

Reply:
you can use suppliers to achieve this , it looks a bit shorter and nicer :) void error(String message, Supplier<?>... paramSuppliers)